Adapt TRAP simulator spec for new digit format

martenole opened the pull request:
- New digits are created for shared pads, so the filling of the TRAP chips doesn't need to take that into account

TODOs:
- add correct scaling factors for dy and y
- check if sorting of the digits is still needed
- MC labels not yet added
- remove MC->raw conversion part?

martenole committed:
Create additional digits for ADC channels connected to shared pads

martenole committed:
Adapt TRAP simulator spec for new digit format

- New digits are created for shared pads, so the filling of the TRAP chips doesn't need to take that into account
- TODOs:
- add correct scaling factors for dy and y
- check if sorting of the digits is still needed
- MC labels not yet added
- remove MC->raw conversion part?

martenole committed:
Fix offset and slope in new Run 3 tracklet format

- cleanup of Tracklet64 class

tdietel commented on DataFormats/Detectors/TRD/include/DataFormatsTRD/Tracklet64.h:
```diff
@@ -63,8 +63,8 @@ class Tracklet64
-  uint64_t getColumn() const { return ((mtrackletWord & colmask) >> colbs); };       // in units of
+  uint64_t getPadRow() const { return ((mtrackletWord & padrowmask) >> padrowbs); }; // pad row number [0..15]
+  uint64_t getColumn() const { return ((mtrackletWord & colmask) >> colbs); };       // column refers to MCM position in column direction on readout board [0..3]
   uint64_t getPosition() const { return ((mtrackletWord & posmask) >> posbs); };     // in units of 0.02 pads [10bits] .. -10.22 to 10.22
```
I noticed taht this comment is outdated.

martenole replied:
I updated all comments in the Tracklet64 header

tdietel commented on Detectors/TRD/macros/convertRun2ToRun3Digits.C:
```diff
@@ -186,9 +188,19 @@ void convertRun2ToRun3Digits(TString qaOutPath = "",
+              isSharedRight = true;
+            } else if (col % NCOLMCM == 17) {
+              isSharedLeft = true;
+            }
```
I am not sure if I understand the code correctly, but it looks like you duplicate the entries from the run 2 digits manually. There should be a function, maybe in the AliTRDdigitsArray, that gives you access to the "hidden" ADC channels directly, without the need to duplicate.

martenole replied:
Yes, this is simply duplicating the digits manually. I can also check later on how to get the data properly. At the moment I am still struggling with the new tracklet format.

martenole replied:
This is still not done, but I would do it in a separate PR later, I think it is not strictly required

tdietel reviewed (commented):
I have looked through the PR somewhat superficially. I made one comment, but that would not stop the merge. 

Do you want to merge multiple steps, or everything at once when it's done?

martenole commented:
@tdietel I want to fix first the offset and position of the tracklets in the run 3 format. So far I could only reproduce the run 2 results using the old run 2 format. But now that I have this as a reference I can debug the new tracklet format more easily. But it is a lot fiddling with bits, signed and unsigned values and this takes some quite time. Hope that I have it done later today

martenole commented:
Hi @bazinski and @tdietel 
from my side this PR can be merged once the checks are through. Offset and slope are fine in the run 3 tracklet format. What still needs to be done is MC labels and the charge windows also don't look right yet, but this is then for a later PR. For the global tracking this should already be something to play with.
Could you please take a look? I put the deflection correction in which was also there in Run 2 (see additional comment in the code). Not sure if it will still be done in the TRAP as well or if we should take it out.

martenole commented on Detectors/TRD/simulation/src/TrapSimulator.cxx:
```diff
@@ -2007,14 +1994,16 @@ void TrapSimulator::fitTracklet()
-      //slope = slope;//this is here as a reference for what was done in run2 //* ndrift) >> ndriftDp) + deflCorr;
-      auto oldpos = position;
+
+      slope = slope + ((deflCorr / ndrift) << ndriftDp);
```
This is the deflection correction I am not sure about..

bazinski replied:
It can be adjusted or redone later, might be easier to diagnose further down the pipeline?

martenole replied:
Agree, would also leave it in for now

bazinski commented:
@martenole the mc to raw stuff can be taken out, so the stuff about the hcid and the associated index can be removed. Turns out it was wrong in anycase. This all redone in trap2raw. I will comment in the code.

bazinski commented on Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx:
```diff
@@ -286,22 +242,23 @@ void TRDDPLTrapSimulatorTask::run(o2::framework::ProcessingContext& pc)
   auto inputDigits = pc.inputs().get<gsl::span<o2::trd::Digit>>("digitinput");
-  std::vector<o2::trd::Digit> msgDigits(inputDigits.begin(), inputDigits.end());
-  //  auto digits pc.outputs().make<std::vector<o2::trd::Digit>>(Output{"TRD", "TRKDIGITS", 0, Lifetime::Timeframe}, msgDigits.begin(), msgDigits.end());
+  std::vector<o2::trd::Digit> msgDigits(inputDigits.begin(), inputDigits.end()); // TODO: check if sorting the digits is still required. If not, this copy can be removed
```
No it is no longer needs to be sorted. So long as the "double" entry digits is done in the digitizer, maybe for now check that the shared pad digits are in fact appearing twice in 2 mcm/rob pair combinations.
The sorting was to put everything into pad rows. 
However ! I assume you want tracklet data in some or other coherent sorted way, so for that purpose you might want to keep this and but change the sort criterion? Else sort tracklets pre tracking?

martenole replied:
I know the 8 trap chips for a specific row are no longer needed, but I wonder if it not makes sense to limit the number of TRAP chips which we keep in memory. If we fill ~65k `TrapSimulator` objects in parallel that might blow up. But I have not yet tested it.
That is why I kept the 8 (and also the digit sorting by pad row). But one could also make a loop for each chamber with 128 (96) TRAPs, maybe that is better.
Overall I have the impression that the `o2-trd-trap-sim` workflow got slower with my changes :/

bazinski commented on Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx:
```diff
@@ -314,12 +271,10 @@ void TRDDPLTrapSimulatorTask::run(o2::framework::ProcessingContext& pc)
-  //TODO these must be created directly in the output as done at the top of this run method
-  //msgDigitsIndex.reserve(msgDigits.size());
-  LOG(debug) << "Read in msgDigits with size of : " << msgDigits.size() << " labels contain : " << digitMCLabels.getNElements() << " with and index size of  : " << digitMCLabels.getIndexedSize() << " and triggerrecord count of :" << triggerRecords.size();
+  mLinkRecords.reserve(2 * MAXCHAMBER * inputTriggerRecords.size()); // worse case scenario is all links for all events
```
linkrecords is related to the raw data so can go. similarly all the code dependent on it.

bazinski commented on Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx:
```diff
@@ -330,35 +285,17 @@ void TRDDPLTrapSimulatorTask::run(o2::framework::ProcessingContext& pc)
   //sort the digits array TODO refactor this intoa vector index sort and possibly generalise past merely digits.
   auto sortstart = std::chrono::high_resolution_clock::now();
-  for (auto& trig : triggerRecords) {
+  // TODO check if sorting is still needed
```
no its not needed in its current form. However some form of sorted maybe needed given the comment above.

bazinski commented on Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx:
```diff
@@ -376,143 +313,129 @@ void TRDDPLTrapSimulatorTask::run(o2::framework::ProcessingContext& pc)
+
     //Are we on a new half chamber ?
-    if (mOldHalfChamberLinkId != LinkRecord::getHalfChamberLinkId(detector, rob)) {
+    if (mOldHalfChamberLinkId != LinkRecord::getHalfChamberLinkId(digit->getDetector(), digit->getROB())) {
```
all of this if statement can go. its logic has been moved to raw data gen.

bazinski replied:
This also holds for lines:
554,555,556 
and the workflow can lose the rawwriter device.

martenole replied:
Removed the raw writer

bazinski commented on Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx:
```diff
@@ -523,58 +446,80 @@ void TRDDPLTrapSimulatorTask::run(o2::framework::ProcessingContext& pc)
+
+    trapTracklets = mTrapSimulator[trapcounter].getTrackletArray64(); //TODO remove the copy and send the Accumulated array into the Trapsimulator
+    auto trapLabels = mTrapSimulator[trapcounter].getTrackletLabels();
+    if (!mNewTrackletHCHeaderHasBeenWritten && trapTracklets.size() != 0) { // take account of the case where we have data in the trapchip adc but no tracklets OS: Is this a TODO?
```
not a todo, an explanation.

bazinski replied:
but entire if and else can go. its now buried in raw generation.i.e. up to line 511

martenole replied:
Well, the comment is confusing with `if trapTracklets.size() !=0` and it says to take account of the case with no tracklets. But ok, will remove the block

bazinski replied:
the header still has to be added to the index even if nothing there.

martenole replied:
So should it be `trapTracklets.size() == 0` then? But anyhow, the block is removed now

bazinski reviewed (changes requested):
the storage of 8 trapsimulators and the looping over 8 was only needed when the incoming pads were sorted into pad rows and then the loop was for an entire pad row. Now that digits are different this is nolonger needed but I suppose can be left in for now ? Although its not clear what defines the 8 now that its not a padrow? Or do you intend to leave the sorting digits into a padrow?

martenole committed:
Remove raw data stuff from TRAPSimSpec

martenole commented:
@bazinski I removed all the raw stuff and rewrote the digits loop. Most of the processing time (>95%) is spent in the actual TRAP processing, so there is not much performance to be gained I think. There are 128 TRAP objects now instead of 8, so the processing is done on the level of chambers instead of pad rows.
Could you have a look again and shout if I have removed too much?

martenole committed:
Fix digit loop counting for trigger records

martenole committed:
Remove more obsolete includes and variables

bazinski commented on Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx:
```diff
@@ -622,11 +330,7 @@ o2::framework::DataProcessorSpec getTRDTrapSimulatorSpec()
                            Outputs{OutputSpec{"TRD", "TRACKLETS", 0, Lifetime::Timeframe}, // this is the 64 tracklet words
                                    OutputSpec{"TRD", "TRKTRGRD", 0, Lifetime::Timeframe},
                                    /*OutputSpec{"TRD", "TRKDIGITS", 0, Lifetime::Timeframe},*/
```
Can you put the digits back and their associated index? Possibly an option to enable/disable it.

martenole replied:
Will come with a following PR

bazinski reviewed (dismissed):
I think all the raw stuff is out. Chambers seem more logical way of organising it.

tdietel reviewed (dismissed):
Looks ok as far as I can tell. Let's merge it and see.

martenole committed:
More cleanup for Tracklet64

martenole committed:
Add missing const in getter function

martenole committed:
Add option to manually take care of shared pads

martenole commented:
Hi @davidrohr,
could you please merge this when you get a chance?
Cheers,
Ole
